bpf: Introduce wildcard service entries to drop traffic towards unknown protocol/port combinations

ajmmm opened the pull request:
PR to resolve a forwarding bug in data path.

This work is derived from work undertaken by Mikael Knutsson <mikael.knutsson@gmail.com> so has been credited in the relevant commit logs.

**Summary of issue**

Given a configured LoadBalancer or Cluster IP on a specific L4 destination port and protocol, ingress traffic towards the IP on an alternate destination port or protocol will not match eBPF service map lookup, so traffic be punted back to the network. Where the upstream network re-delivers the traffic back to Cilium, a packet loop is formed.

**Summary of Changes**
- Introduces the concept of "wildcard" service entries in eBPF Service Maps, being a service entry with a zero destination and protocol field.
- LoadBalancer and ClusterIP frontend services with external scope will be cross-referenced to a wildcard entry:
  - The first reference will trigger addition of wildcard entry in eBPF
  - The last reference deleted will trigger removal of wildcard entry from eBPF
- Data path behaviour modified on traffic ingress. Where service lookup fails, an additional lookup is undertaken with zero destination and protocol. On successful match to such a wildcard entry, traffic is dropped.
- Removal of data path logic that maintained compatibility with eBPF maps programmed by older versions of Cilium. Protocol differentiation was deprecated in 1.17, and automatic conversion of undifferentiated protocol was added in 1.18.
- Appropriate unit test updates for the new logic
- Minor typo fixes in unit test logic

Release Notes:

```release-note
Introduce wildcard service entries to ensure traffic towards a LoadBalancer and ClusterIPs with an unknown protocol/port combination is dropped by the data path, rather than being forwarded back to the network.
```

ajmmm committed:
bpf: Remove redundant map key fixups in lbX_lookup_service() failure paths.

Previously, lbX_lookup_service() would zeroise the key protocol field,
such that callers would	need to	reinstate the key protocol field if their
logic re-used the key.

Following the previous commit, the lbX_lookup_service()	routines will now
only zeroise the key protocol (and dport) fields if an external	wild-card
entry is returned. Thus, the restoration of key values can be removed.

This commit also removes lbX_key_set_protocol() helpers as they became
broadly unnecessary.

Signed-off-by: Alasdair McWilliam <alasdair.mcwilliam@isovalent.com>

ajmmm committed:
loadbalancer: Extend reconciler nodePortAddrKey to include L4 Protocol

This change extends the nodePortAddrKey structure to include the L4 Protocol
number of an address, such that addresses with differing addresses will be
better distributed in the nodePortAddrByPort map.

This required the introduction of a helper routine to translate the L4Addr
protocol field into the real IANA number.

Signed-off-by: Alasdair McWilliam <alasdair.mcwilliam@isovalent.com>

ajmmm committed:
loadbalancer: Update unit tests

Update the unit tests for loadbalancer and ciliumenvoyconfig to include
wildcard entries.

Signed-off-by: Alasdair McWilliam <alasdair.mcwilliam@isovalent.com>

ajmmm committed:
bpf: tests: Remove redundant function

Remove lb_v4_add_mixed_proto_service_with_flags() as it no longer makes sense.

Signed-off-by: Alasdair McWilliam <alasdair.mcwilliam@isovalent.com>

joamaki commented on pkg/loadbalancer/reconciler/bpf_reconciler.go:
```diff
@@ -1202,6 +1302,151 @@ func (ops *BPFOps) updateMaglev(fe *loadbalancer.Frontend, feID loadbalancer.Ser
+	// If the Frontend Address is already exists in the wildcardAddrs set, we take
+	// no action because the FE is already be mapped to this wildcard service entry.
+
+	if ops.wildcardAddrs[addr].Has(fe.Address) {
```
do a single lookup into  `ops.wildcardAddrs` instead of many to avoid unnecessary work:
```
addrs := ops.wildcardAddrs[addr]
if addrs == nil {
  addrs = sets.New..
  ops.wildcardAddrs[addr] = addrs
}

if addrs.Has(fe.Address) { ... }
if addrs.Len() == 0 { ...}
addrs.Insert(fe.Address)

```

hashing + hash table lookups are pretty fast, but it does add up.

joamaki commented on pkg/loadbalancer/reconciler/bpf_reconciler.go:
```diff
@@ -1202,6 +1302,151 @@ func (ops *BPFOps) updateMaglev(fe *loadbalancer.Frontend, feID loadbalancer.Ser
+
+	addr := fe.Address.AddrCluster.Addr()
+
+	if ops.wildcardAddrs[addr] == nil {
```
Same here, avoid the multiple hash table lookups

joamaki commented on pkg/loadbalancer/reconciler/bpf_reconciler.go:
```diff
@@ -475,18 +501,58 @@ func (ops *BPFOps) deleteFrontend(fe *loadbalancer.Frontend) error {
+			// We only add this entry into toDelete if it has nothing in. Otherwise, we may
+			// end up deleting a wildcard who still has active parent entries.
+
+			if ops.wildcardAddrs[wildAddr] == nil || ops.wildcardAddrs[wildAddr].Len() == 0 {
```
Avoid multiple hash table lookups here too.

joamaki commented on pkg/loadbalancer/reconciler/bpf_reconciler.go:
```diff
@@ -995,6 +1064,17 @@ func (ops *BPFOps) updateFrontend(fe *loadbalancer.Frontend) error {
+	// Upsert wildcard entries such that the data path will have a service entry for any
+	// LoadBalancer or Cluster IP that receives traffic for an unknown protocol/port combination.
+
+	ops.log.Debug("Update wildcard", logfields.ID, feID)
```
I don't think this debug log is that useful considering `upsertWildcard` logs if it adds the entry

joamaki commented on pkg/loadbalancer/reconciler/bpf_reconciler.go:
```diff
@@ -1202,6 +1302,151 @@ func (ops *BPFOps) updateMaglev(fe *loadbalancer.Frontend, feID loadbalancer.Ser
+//
+// See upsertWildcard() for semantics. This does the reverse.
+//
+// @param[in]    fe    Pointer to a Frontend object
```
I've never seen these sort of java-style comments in Go code or in Cilium. Let's not add them and stick to normal conventions instead.

joamaki commented on pkg/loadbalancer/reconciler/bpf_reconciler.go:
```diff
@@ -1202,6 +1302,151 @@ func (ops *BPFOps) updateMaglev(fe *loadbalancer.Frontend, feID loadbalancer.Ser
+// This routine will have no effect if the frontend type is not LoadBalancer or ClusterIP,
+// or if the Frontend is already present in the wildcardAddrs map set.
+//
+// @param[in]    fe    Pointer to a Frontend object
```
Remove these java-style comments. Only document parameters and return values if there's something non-obvious to say about them ("Pointer to a Frontend object" is obvious from `fe *Frontend` and it's obvious that if you have `error` as return type that `nil` is success). I would go as far as not even add a doc comment to a method if it's internal and it's obvious from its name and types what it does, though for this one it makes a lot of sense.

ajmmm replied:
I've removed the `@param` style tags - but let me know if you would prefer I remove the whole comment.

joamaki commented on pkg/loadbalancer/reconciler/bpf_reconciler.go:
```diff
@@ -1058,6 +1139,20 @@ func (ops *BPFOps) useMaglev(fe *loadbalancer.Frontend) bool {
 	}
 }
 
+// Identify if a Frontend should invoke wildcard service entries.
```
Does this method actually need a comment? Seems obvious from the method name.

joamaki commented on pkg/loadbalancer/reconciler/bpf_reconciler.go:
```diff
@@ -150,6 +158,10 @@ type BPFOps struct {
 	// This is used when updating to remove orphans.
 	prevSourceRanges map[loadbalancer.L3n4Addr]sets.Set[netip.Prefix]
 
+	// wildcardAddrs maps an IP Address to a set of parent LoadBalancer or ClusterIP
```
Mention here that it's a frontend IP address and also why we maintain these.

joamaki commented on pkg/loadbalancer/reconciler/bpf_reconciler.go:
```diff
@@ -1024,6 +1106,7 @@ func (ops *BPFOps) updateFrontend(fe *loadbalancer.Frontend) error {
 
 	// Finally update the new references. This makes sure any failures reconciling the service slots
 	// above can be retried and entries are not leaked.
+
```
What's up with the extra newlines between comment and code? It is rather unconventional.

ajmmm replied:
A general preference to avoid large blocks of text building up and becoming difficult to separate. But I take your point! Didn't think it was an issue - didn't get mentioned on a previous PR and was thanked!

The extra whitespace has been removed.

dylandreimerink commented on bpf/lib/nodeport.h:
```diff
@@ -1460,6 +1460,17 @@ static __always_inline int nodeport_lb6(struct __ctx_buff *ctx,
+		 */
+		if (key.dport == LB_SVC_WILDCARD_DPORT) {
+			ctx_set_xfer(ctx, XFER_PKT_NO_SVC);
+			return send_drop_notify_error(ctx, src_sec_identity, DROP_NO_SERVICE,
```
Doing a drop notify here will cause a double log. This is a function called by others. It is typically the job of the first function / entry point function to actually do the `send_drop_notify_...`. 

In this case, if you go look at the call stack:
- nodeport_lb6
  - handle_ipv6 @ bpf_host.c
    - [tail_handle_ipv6 @ bpf_host.c](https://github.com/cilium/cilium/blob/129e2182e4f187338b3717a6c6fec6a16b0cb843/bpf/bpf_host.c#L461)
  - handle_ipv6 @ bpf_overlay.c
    - [tail_handle_ipv6 @ bpf_overlay.c](https://github.com/cilium/cilium/blob/129e2182e4f187338b3717a6c6fec6a16b0cb843/bpf/bpf_overlay.c#L240)
  - handle_ipv6 @ bpf_wireguard.c
    - [tail_handle_ipv6 @ bpf_wireguard.c](https://github.com/cilium/cilium/blob/129e2182e4f187338b3717a6c6fec6a16b0cb843/bpf/bpf_wireguard.c#L124)
  - [tail_lb_ipv6 @ bpf_xdp.c](https://github.com/cilium/cilium/blob/129e2182e4f187338b3717a6c6fec6a16b0cb843/bpf/bpf_xdp.c#L285)

You will see that the base function sends the notify with the return value, and often an extended error code `ext_err`. So if you also do one in this function, it will send one message here, but also one when we go back to the outer function with `CTX_ACT_DROP` as reason (since that will be the return value).

So I think we want the code here to be:
```
			ctx_set_xfer(ctx, XFER_PKT_NO_SVC);
                        // *ext_err = additional code to distinguish from other DROP_NO_SERVICE usages // optional
			return DROP_NO_SERVICE;
```

I did notice that the outer send_drop_notify_... does not pass `METRIC_SERVICE` as direction, they tend to use `METRIC_INGRESS` instead. So if that is important, you may need to plumb the direction value in via arguments as well.

ajmmm replied:
I'd been meaning to ask about that actually - thank you!
I'll revise shortly

ajmmm replied:
Only seems to be a few use cases where passing `ext_err` is useful, so I've modified as suggested to just return `DROP_NO_SERVICE`. That error is descriptive of the problem. Furthermore, on the basis this is by definition ingress traffic I think the ingress direction is acceptable vs. the complexity of having an extra argument.

dylandreimerink commented on bpf/lib/nodeport.h:
```diff
@@ -2797,6 +2808,17 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
+		 */
+		if (key.dport == LB_SVC_WILDCARD_DPORT) {
+			ctx_set_xfer(ctx, XFER_PKT_NO_SVC);
+			return send_drop_notify_error(ctx, src_sec_identity, DROP_NO_SERVICE,
```
Same here, see other comment for details.

ajmmm replied:
Feedback above

dylandreimerink reviewed (changes requested):
This looks great overall. One point of feedback though.

ajmmm committed:
bpf: lb: Remove unnecessary service map lookups in __lbXX_lookup_service().

Previously, if the lookup of a service derived from ingress 5-tuple results
in no match, we undertake an additional lookup with IPPROTO_ANY. This was
necessary to maintain backwards compatibility with earlier versions of Cilium
that did not support protocol differentiation when programming the data path.

Cilium 1.18 introduced logic to migrate IPPROTO_ANY service/backend BPF map
entries to an appropriate protocol automatically, such that the additional
lookups are no longer required.

Appropriate tests updated to include explicit lookups for IPPROTO_TCP,
IPPROTO_UDP, and mismatched protocol checks.

Signed-off-by: Alasdair McWilliam <alasdair.mcwilliam@isovalent.com>

dylandreimerink reviewed (approved):
Thanks for all of the work that has gone into this! Changes look good

ajmmm committed:
bpf: lb: Add wildcard drop logic for LB VIPs on unknown port/protocol.

This change modifies the behaviour of lbXX_lookup_service() such that if an external
service lookup does not find an entry, we undertake an additional lookup with DPort=0
and Protocol=0. This will facilitate a wildcard lookup that, when programmed with a
suitable wildcard entry by the control plane, will trigger the data path to drop
unmatched ingress traffic.

This change modifies the semantics of said routines such that they no longer propagate
changes to the key protocol field, unless a wildcard service entry was found in the
external scope, and returned. In this case, the dport and protocol field will be zero.

Signed-off-by: Alasdair McWilliam <alasdair.mcwilliam@isovalent.com>
Signed-off-by: Mikael Knutsson <mikael.knutsson@gmail.com>

ajmmm committed:
bpf: tests: Add test cases for service wildcard drop entries.

This set of tests verifies that if we receive a packet towards a VIP on either
an unknown destination port, or unknown protocol, the data plane will correctly
drop if a wildcard entry is present in the service map.

Signed-off-by: Alasdair McWilliam <alasdair.mcwilliam@isovalent.com>

ajmmm committed:
loadbalancer: Add logic to program wildcard (drop) entries in eBPF.

Introduce logic in bpf_reconciler to map frontend service entries to wildcard
service entries, and program them in the data path, for LoadBalancer and ClusterIP
frontends of external scope. This will ensure the data path drops traffic towards
unknown destination ports and/or protocols.

Wildcard state is tracked as a map (indexed by IP Address) that contains a slice
of Frontend ServiceIDs. The first entry that requires a wildcard will trigger
programming of the data path entry. The last entry that depends on a wildcard
entry to be removed will trigger removal of the wildcard entry.

Signed-off-by: Alasdair McWilliam <alasdair.mcwilliam@isovalent.com>
Signed-off-by: Mikael Knutsson <mikael.knutsson@gmail.com>

borkmann reviewed (approved):
(lgtm, still needs a rebase)

borkmann commented on bpf/lib/lb.h:
```diff
@@ -684,37 +677,54 @@ lb6_to_lb4_service(const struct lb6_service *svc __maybe_unused)
 	key->backend_slot = 0;
 
+	/* External scope lookup */
+
```
small nit: Do we need this extra comment?

borkmann commented on bpf/lib/lb.h:
```diff
@@ -684,37 +677,54 @@ lb6_to_lb4_service(const struct lb6_service *svc __maybe_unused)
+	if (!svc) {
+		/* If no service exists, look for a wildcard entry, which should be
+		 * programmed with dport = WILDCARD_DPORT and proto = WILDCARD_PROTO.
+		 */
```
same here, this basically comments what the code does which is redundant. I think we should only have comments on why we do something to provide additional context.

borkmann commented on bpf/lib/lb.h:
```diff
@@ -684,37 +677,54 @@ lb6_to_lb4_service(const struct lb6_service *svc __maybe_unused)
+		if (svc)
+			return svc;
+
+		/* Ensure we restore the original key values */
```
same here

borkmann commented on bpf/lib/lb.h:
```diff
@@ -684,37 +677,54 @@ lb6_to_lb4_service(const struct lb6_service *svc __maybe_unused)
+
+		/* If we have no external scope for this, it's safe to return NULL here
+		 * because there can't be an internal scope. (For now.)
+		 */
```
This one looks good and is reasonable to stay.

borkmann commented on bpf/lib/lb.h:
```diff
@@ -1415,37 +1412,61 @@ lb4_to_lb6_service(const struct lb4_service *svc __maybe_unused)
+ *
+ * Where scope_switch is TRUE, an external non-wildcard lookup must be found
+ * before search can progress to an internal lookup.
+ */
```
nit: same here.. I would probably remove it so that this doesn't go out of sync with the actual code

borkmann commented on bpf/lib/lb.h:
```diff
@@ -1415,37 +1412,61 @@ lb4_to_lb6_service(const struct lb4_service *svc __maybe_unused)
+		key->dport = LB_SVC_WILDCARD_DPORT;
+		key->proto = LB_SVC_WILDCARD_PROTO;
+
+		svc = __lb4_lookup_service(key);
```
I think one small comment which would make sense here imho would be to state that these service entries have zero backends thus later in the code they trigger a drop.. to better understand why we return these svc?

Also the wildcard naming is a bit unfortunate since in socket lb we also have wildcard lookup but its different and for a different purpose. Probably ok to leave as-is but something to watch out for when touching the code.

borkmann commented on bpf/lib/lb.h:
```diff
@@ -1415,37 +1412,61 @@ lb4_to_lb6_service(const struct lb4_service *svc __maybe_unused)
 		return svc;
 
-	key->proto = orig_proto;
+	/* Perform internal scope lookup */
```
same nit

borkmann reviewed (approved):
bpf side lgtm! some small nits when you rebase anyway.

julianwiedmann commented on bpf/lib/nodeport.h:
```diff
@@ -2839,6 +2849,16 @@ static __always_inline int nodeport_lb4(struct __ctx_buff *ctx,
+		 * delivered to is in a loop.
+		 */
+		if (key.dport == LB_SVC_WILDCARD_DPORT) {
+			ctx_set_xfer(ctx, XFER_PKT_NO_SVC);
```
@ajmmm :wave: since we drop the packet, this line isn't actually needed right?

`XFER_PKT_NO_SVC` is what we set from the XDP layer when we already handled the SVC processing. And so once the packet reaches the TC program, it doesn't need another sequence of SVC processing.

If that makes sense, I'll clean it up as part of a PR I have pending ...

ajmmm replied:
Good catch :) Thanks for pointing it out !!!

julianwiedmann replied:
-> https://github.com/cilium/cilium/pull/42054
